zhmcclient_mock/_hmc.py

```diff
@@ -0,0 +1,1006 @@
+# limitations under the License.
+
+"""
+The `zhmcclient_mock` package provides a faked HMC with all resources that are
+relevant for the `zhmcclient` package. The faked HMC is implemented as a
+local Python object and maintains its resource state across operations.
+"""
+
+from __future__ import absolute_import
+
+try:
+    from collections import OrderedDict
+except ImportError:
+    from ordereddict import OrderedDict
+import six
+# from six.moves.urllib.parse import urlparse, parse_qsl
+
+# TODO: Move the resources into their own files.
+
+__all__ = ['FakedBaseResource', 'FakedBaseManager', 'FakedHmc',
+           'FakedActivationProfileManager', 'FakedActivationProfile',
+           'FakedAdapterManager', 'FakedAdapter',
+           'FakedCpcManager', 'FakedCpc',
+           'FakedHbaManager', 'FakedHba',
+           'FakedLparManager', 'FakedLpar',
+           'FakedNicManager', 'FakedNic',
+           'FakedPartitionManager', 'FakedPartition',
+           'FakedPortManager', 'FakedPort',
+           'FakedVirtualFunctionManager', 'FakedVirtualFunction',
+           'FakedVirtualSwitchManager', 'FakedVirtualSwitch',
+           ]
+
+
+class FakedBaseResource(object):
+    """
+    A base class for faked resource classes in the faked HMC.
+    """
+
+    def __init__(self, manager, properties):
+        self._manager = manager
+        self._properties = properties
+
+        if self.manager.oid_prop not in self.properties:
+            new_oid = self.manager._new_oid()
+            self.properties[self.manager.oid_prop] = new_oid
+        self._oid = self.properties[self.manager.oid_prop]
+
+        if self.manager.uri_prop not in self.properties:
+            new_uri = self.manager.base_uri + '/' + self.oid
+            self.properties[self.manager.uri_prop] = new_uri
+        self._uri = self.properties[self.manager.uri_prop]
+
+    @property
+    def manager(self):
+        """
+        The manager for this resource (a derived class of
+        :class:`~zhmcclient_mock.FakedBaseManager`).
+        """
+        return self._manager
+
+    @property
+    def properties(self):
+        """
+        The properties of this resource (a dictionary).
+        """
+        return self._properties
+
+    @property
+    def oid(self):
+        """
+        The object ID (property 'object-id' or 'element-id') of this resource.
+        """
+        return self._oid
+
+    @property
+    def uri(self):
+        """
+        The object URI (property 'object-uri' or 'element-uri') of this
+        resource.
+        """
+        return self._uri
+
+    def update(self, properties):
+        """
+        update the properties of this resource.
+
+        Parameters:
+
+          properties (dict):
+            Resource properties to be updated. Any other properties remain
+            unchanged.
+        """
+        self.properties.update(properties)
+
+
+class FakedBaseManager(object):
```

Comment:
Creation and deletion of e.g. partitions is supported. Again, the testcase uses the external zhmcclient API, just with a FakedSession instead of the real Session.